### PR TITLE
async-coap-uri: improve ParseError

### PR DESCRIPTION
**Pull request by Luro02**

### CODE REVIEW

File: async-coap-uri/src/escape/unescape_uri.rs [hunk 1]
```diff
@@ -131,7 +134,7 @@ impl<'a> UnescapeUri<'a> {
     /// If the string cannot be decoded losslessly, then the location of the encoding
     /// error is returned as an `Err`.
```
Comment (darconeous) [inline async-coap-uri/src/escape/unescape_uri.rs:135]: This comment should be updated.

Comment (Luro02) [reply async-coap-uri/src/escape/unescape_uri.rs]: I fixed that :)

File: async-coap-uri/src/escape/unescape_uri.rs [hunk 2]
```diff
@@ -202,6 +199,97 @@ impl<'a> UnescapeUri<'a> {
...
+            Self::InvalidByte(b) => write!(f, "not a valid utf8 character `{}`", b),
+            Self::UnexpectedChar(c) => write!(f, "unexpected char `{:?}`", c),
+        }
+    }
+}
+
+#[cfg(feature = "std")]
+impl ::std::error::Error for UnescapeError {}
+
 impl<'a> Iterator for UnescapeUri<'a> {
```
Comment (Luro02) [inline async-coap-uri/src/escape/unescape_uri.rs:293]: Before merging could you please look through this implementation for mistakes or error messages that can be improved?
There were some parts in this function that I did not understand, and I think there are some errors returned that are wrong or could be improved :)

File: async-coap-uri/src/escape/unescape_uri.rs [hunk 3]
```diff
@@ -311,16 +406,20 @@ impl<'a> Iterator for UnescapeUri<'a> {
...
+
                         if utf8_len >= UTF8_CHAR_WIDTH[utf8_buf[0] as usize] as usize {
                             if let Ok(utf8_str) = from_utf8(&utf8_buf[..utf8_len]) {
                                 return Some(utf8_str.chars().next().unwrap());
                             } else {
-                                self.had_error = true;
+                                self.had_error = Some(UnescapeError::InvalidUtf8 {
+                                    buf: utf8_buf,
+                                    len: utf8_len,
+                                });
```
Comment (Luro02) [inline async-coap-uri/src/escape/unescape_uri.rs:417]: This is one of the errors that I am not sure about

File: async-coap-uri/src/escape/unescape_uri.rs [hunk 4]
```diff
@@ -311,16 +406,20 @@ impl<'a> Iterator for UnescapeUri<'a> {
...
                                 return Some(REPLACEMENT_CHARACTER);
                             }
                         }
                     } else if utf8_len != 0 {
-                        self.had_error = true;
+                        self.had_error = Some(UnescapeError::InvalidByte(decoded));
```
Comment (Luro02) [inline async-coap-uri/src/escape/unescape_uri.rs:422]: This is another one, where I am not sure what it is supposed to mean/what the conditions were to reach this error?

Comment (darconeous) [reply async-coap-uri/src/escape/unescape_uri.rs:422]: We reach this condition when we have reach a non-UTF8 character when we are in the process of parsing a UTF8 sequence. Take, for example, the emoji `🐙`: it has a URI-encoded sequence of `%f0%9f%90%99`. This check is to make sure we properly error out if we encounter a sequence like `%f0%9f%90%20`. That is an invalid sequence.

File: async-coap-uri/src/escape/unescape_uri.rs (hunk 3 above)
Comment (darconeous) [reply async-coap-uri/src/escape/unescape_uri.rs:417]: This error happens when we have received all of the UTF8 bytes that were indicated by the first byte, but the resulting sequence doesn't decode properly into a valid UTF character. For example, `%f0%9f%90%f0` contains four UTF8 byte codes (they all have the MSB set), but together they do not decode properly.

File: async-coap-uri/src/escape/unescape_uri.rs [hunk 5]
```diff
@@ -77,23 +79,24 @@ impl<'a> From<UnescapeUri<'a>> for Cow<'a, str> {
 impl<'a> FusedIterator for UnescapeUri<'a> {}
 
 impl<'a> UnescapeUri<'a> {
-    /// Returns the location of the first encountered encoding error, if any.
-    pub fn first_error(&self) -> Option<usize> {
+    /// Returns the first encountered encoding error, if any.
+    pub fn first_error(&self) -> Option<DecodingError> {
```
Comment (darconeous) [inline async-coap-uri/src/escape/unescape_uri.rs:83]: Honestly, having the index of the first encoding error is more useful than what the specific error is, IMHO.

File: async-coap-uri/src/error.rs [hunk 6]
```diff
@@ -61,6 +80,75 @@ impl ParseError {
...
+            span: Some(error.index..error.index + 1),
+            desc: ParseErrorKind::EncodingError(error),
+        }
+    }
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
+enum ParseErrorKind {
+    /// Bad percent encoding or illegal characters
+    EncodingError(crate::escape::DecodingError),
```
Comment (darconeous) [inline async-coap-uri/src/error.rs:108]: I get why this is named `EncodingError`, but it does feel a little weird that the variant named `EncodingError` contains a single variable with a type named `DecodingError`.

File: async-coap-uri/src/escape/unescape_uri.rs [hunk 7]
```diff
@@ -253,17 +345,19 @@ impl<'a> Iterator for UnescapeUri<'a> {
                         }
                         None => {
                             self.iter_index -= 1;
-                            self.had_error = true;
+                            self.had_error = Some(UnescapeError::MissingChar);
```
Comment (darconeous) [inline async-coap-uri/src/escape/unescape_uri.rs:348]: "IncompleteEscape", perhaps?

**Review by darconeous — APPROVED** [review]

**Pull request merged**